crypto/src/lib.rs

```diff
@@ -340,17 +366,113 @@ impl PublicKey {
         let payload = hex_decode(payload)?;
         let payload = ConstVec::new(payload);
 
-        // NOTE: PrivateKey does some validation by generating a public key from the provided bytes
-        // we can't really do this for PublicKey
-        // this can be solved if the keys used here would be actually aware of the underlying crypto primitive types
-        // instead of just being raw bytes
-        Ok(Self {
-            digest_function,
-            payload,
+        Self::from_raw(digest_function, &payload)
+    }
+
+    /// Get the digest function of the public key
+    pub fn digest_function(&self) -> Algorithm {
+        match self {
+            Self::Ed25519(_) => Algorithm::Ed25519,
+            Self::Secp256k1(_) => Algorithm::Secp256k1,
+            Self::BlsNormal(_) => Algorithm::BlsNormal,
+            Self::BlsSmall(_) => Algorithm::BlsSmall,
+        }
+    }
+}
+
+#[cfg(not(feature = "ffi_import"))]
+impl fmt::Debug for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple(self.digest_function().as_static_str())
+            .field(&self.normalize())
+            .finish()
+    }
+}
+
+#[cfg(not(feature = "ffi_import"))]
+impl fmt::Display for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&self.normalize())
+    }
+}
+
+#[cfg(not(feature = "ffi_import"))]
+impl core::hash::Hash for PublicKey {
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+        (self.digest_function(), self.payload()).hash(state)
+    }
+}
+
+impl PartialOrd for PublicKey {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for PublicKey {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        (self.digest_function(), self.payload()).cmp(&(other.digest_function(), other.payload()))
+    }
+}
+
+#[cfg(not(feature = "ffi_import"))]
+impl Encode for PublicKey {
+    fn size_hint(&self) -> usize {
+        self.payload().len() + 1
```

Comment:
I don't think this size hint is right, the `ConstVec` will be prefixed with length, so it'll always end up at least one byte longer